### PR TITLE
chore: Remove the subsumption inference used for label polymorphism

### LINKED ISSUE #4740: Analyze the impact of of using polymorphic labels in the standard library
Polymorphic labels were added in #4388 with a couple of example function signatures that utilize the feature. To be able to change the stable builtins to use polymorphic labels we need to evaluate how large the fallout would be.

DOD:

- Provide alternate type signatures which use polymorphic labels for all builtins where it is useful (see https://github.com/influxdata/flux/issues/4410 for an incomplete list)
- When the label polymorphism feature is enabled do the typechecking with these signatures that use labels
- Evaluate the impact by compiling a representative set of real-world queries to determine the regressions

### PR DESCRIPTION
**Pull request by Marwes**
Removes the old subsumption (subtyping) based label polymorphism in preparation of #4927, #4928 and #4929 . Tests that fail are ignored or removed if they assumed subtyping.

I also added the updated type signatures which are hidden by `@feature` usage so they do not appear in the current standard library.

### CODE REVIEW

**Comment by Marwes** [comment]
Trying to think of a solution without involving sub typing.

I am thinking we compile the standard library with label polymorphism everywhere (so there is no alternate signatures to deal with). When checking queries outside of the standard library we do an extra pass before typechecking that checks walks the semantic graph and detects when a builtin using labels is accessed. For these cases, if we can statically infer that a builtin is called with string literals we use the type signature with labels, if we are unable to determine that a string literal is used we translate the type signature with labels to one without.

```flux
builtin sum : (<-tables: stream[{ A with C: B }], ?column: C = "_value") => stream[{ D with C: B }]
    where A: Record, B: Numeric, C: Label
// translate label variables to string and records containing label fields to an unconstrained record variable
// This should match its original signature
builtin sum : (<-tables: stream[A], ?column: string) => stream[B] where A: Record, B: Record
```

The cases where we can statically know that a string literal is used is simply

```flux
sum(column: "mycolumn")

x = "mycolumn2"
sum(column: x) 
// Gets rewritten to sum(column: "mycolumn2") to match the first case
```

This should give label inference for most cases, except when users implement their own functions that takes a label as argument.

```flux
f = (tables, column) => tables |> sum(column: column) // Defaults to the old string based signature
```

**Comment by Marwes** [comment]
Trying to think of a solution without involving sub typing.

I am thinking we compile the standard library with label polymorphism everywhere (so there is no alternate signatures to deal with). When checking queries outside of the standard library we do an extra pass before typechecking that checks walks the semantic graph and detects when a builtin using labels is accessed. For these cases, if we can statically infer that a builtin is called with string literals we use the type signature with labels, if we are unable to determine that a string literal is used we translate the type signature with labels to one without.

```flux
builtin sum : (<-tables: stream[{ A with C: B }], ?column: C = "_value") => stream[{ D with C: B }]
    where A: Record, B: Numeric, C: Label
// translate label variables to string and records containing label fields to an unconstrained record variable
// This should match its original signature
builtin sum : (<-tables: stream[A], ?column: string) => stream[B] where A: Record, B: Record
```

The cases where we can statically know that a string literal is used is simply

```flux
sum(column: "mycolumn")

x = "mycolumn2"
sum(column: x) 
// Gets rewritten to sum(column: "mycolumn2") to match the first case
```

This should give label inference for most cases, except when users implement their own functions that takes a label as argument.

```flux
f = (tables, column) => tables |> sum(column: column) // Defaults to the old string based signature
```

**Comment by Marwes** [comment]
So I think the most reasonable plan to move forward here is to not have any sort of subtyping between labels and strings.

To make use of labels we add a new syntax (I propose `.mylabel`). The type signatures are left as is to avoid breakage in old scripts however we also provide an option to opt-in to the new label based signatures. This option can be enabled on a script by script basis in the GUI (and also possibly via some sort of `"use script";` annotation in the script itself).

An option comes with the disadvantage that it is easy for people to just not use it so we will want to enable it by default on any new scripts that are created. (Potentially we could also check if old scripts work with label polymorphism and enable the option if they work).

**Comment by Marwes** [comment]
I rebased this PR that original was for #4740 and made it into a removal of subtyping + addition of (disabled) type signatures

**Review by scbrickley — APPROVED** [review]
I don't know if I have the context or the domain knowledge to give this a proper review, but from what I do understand based on this morning's conversation, this looks fine to me.

It might be worth it to get a review from someone who has spent more time in the type system code, though.

File: libflux/flux-core/src/semantic/tests/labels.rs [hunk 1]
```diff
@@ -84,6 +87,7 @@ fn labels_dynamic_string() {
 }
 
 #[test]
+#[ignore]
```
[the remark this answers is not in the document]
Comment (Marwes) [reply libflux/flux-core/src/semantic/tests/labels.rs:90]: Updated https://github.com/influxdata/flux/issues/4927 to mention this test module

**Comment by Marwes** [comment]
Shrank this a bit more by splitting out the stdlib changes to #4976

**Pull request closed without merging**